csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -301,27 +133,259 @@ std::unordered_set<ParallelType> getParallelTypesToPropagate(
   return selected_parallel_types;
 }
 
+// Applies the given transforms on ref to target using the ref2target map.
+void transformLoopDomain(
+    TensorView* target,
+    const TensorView* ref,
+    std::unordered_map<IterDomain*, IterDomain*>& ref2target,
+    const std::unordered_set<IterDomain*>& device_ids,
+    PropagateDirection direction) {
+  // Parallelize based on the ref2target map.
+  for (auto& [ref_id, target_id] : ref2target) {
+    target_id->parallelize(ref_id->getParallelType());
+  }
```

Comment:
```suggestion
```
I suspect this is not necessary or can be avoided with some slight changes. Only leaf IterDomains can be parallelized. We can probably instead set parallel types **after** the replay, which we need to do any way.

Reply:
Done